android/src/main/java/io/branch/rnbranch/RNBranchModule.java

```diff
@@ -211,13 +215,21 @@ private void generateLocalBroadcast(JSONObject referringParams,
         Branch.sessionBuilder(reactActivity).withCallback(referralInitListener).withData(uri).init();
     }
 
+    public static void setIntent(Intent intent, @Nonnull Activity activity) {
+        activity.setIntent(intent);
+        mNewIntent = true;
+    }
+
     /**
      * Notify JavaScript of init session start. This generates an RNBranch.initSessionStart
      * event to JS via the RN native event emitter.
      * @param context a Context for the LocalBroadcastManager
      * @param uri the URI to include in the notification or null
      */
     private static void notifyJSOfInitSessionStart(Context context, Uri uri) {
+        if (!mNewIntent) return;
```

Comment:
What happens to intra-app linking in this scenario? Does it still work?

Reply:
This change only limits 1) the `onOpenStart` call to JS; and 2) the `uri` property in the JS `onOpenComplete` property. You can see it in this branch in the output of `webview_example` (doesn't use the new method); and 2) `testbed_simple` (does use the new method). In each case, I opened a link into the app, the backgrounded and foregrounded the app.

webview_example: onOpenStart called a second time on foreground, though no link opened. Note that `onOpenComplete` does _not_ return the link a second time.
```diff
[Wed Jul 29 2020 16:10:13.240]  LOG      Opening URI null 
[Wed Jul 29 2020 16:10:13.241]  LOG      Branch params: {"+is_first_session":false,"+clicked_branch_link":false,"+rn_cached_initial_event":true}
[Wed Jul 29 2020 16:10:45.226]  LOG      Opening URI https://ms9y.test-app.link/7u2DBzzPu8 
[Wed Jul 29 2020 16:10:45.434]  LOG      Branch params: {"~channel":"Copy","~id":"816737824401657427","$one_time_use":false,"$canonical_identifier":"planet/Earth","$desktop_url":"https://en.wikipedia.org/wiki/Earth","+match_guaranteed":true,"source":"android","~creation_source":2,"$publicly_indexable":"true","$ios_deepview":"branch_default","~referring_link":"https://ms9y.test-app.link/7u2DBzzPu8","+is_first_session":false,"$canonical_url":"https://en.wikipedia.org/wiki/Earth","+clicked_branch_link":true,"~feature":"share","$og_image_url":"https://upload.wikimedia.org/wikipedia/commons/9/97/The_Earth_seen_from_Apollo_17.jpg","$identity_id":"816737567340899405","+click_timestamp":1596064245,"$og_title":"Earth"}
[Wed Jul 29 2020 16:10:46.765]  LOG      Created Branch Universal Object and logged standard view item event.
+[Wed Jul 29 2020 16:10:55.440]  LOG      Opening URI https://ms9y.test-app.link/7u2DBzzPu8 +
[Wed Jul 29 2020 16:10:55.490]  LOG      Branch params: {"+is_first_session":false,"+clicked_branch_link":false}
```

testbed_simple: onOpenStart only called once. Note also that `uri` is `null` here. Though not logged above by webview_example, it is non-null when the app comes back from the foreground with an empty deferred deep link check.
```diff
[Wed Jul 29 2020 16:16:48.970]  LOG      Branch opening URI null [cached]
[Wed Jul 29 2020 16:16:48.970]  INFO     Received link response from Branch for null
[Wed Jul 29 2020 16:16:48.980]  LOG      params: {"+is_first_session":false,"+clicked_branch_link":false,"+rn_cached_initial_event":true}
[Wed Jul 29 2020 16:16:55.262]  LOG      Branch opening URI https://zjef.test-app.link/n4efBZnSu8 
[Wed Jul 29 2020 16:16:55.956]  INFO     Received link response from Branch for https://zjef.test-app.link/n4efBZnSu8
[Wed Jul 29 2020 16:16:55.957]  LOG      params: {"~channel":"Copy","~id":"816747731456192789","$one_time_use":false,"$canonical_identifier":"abc","+match_guaranteed":true,"source":"android","~creation_source":2,"$publicly_indexable":"true","~referring_link":"https://zjef.test-app.link/n4efBZnSu8","+is_first_session":false,"+clicked_branch_link":true,"~feature":"share","$identity_id":"815317995815593579","+click_timestamp":1596064615,"$og_title":"wallo"}
[Wed Jul 29 2020 16:17:00.959]  INFO     Received link response from Branch for null
[Wed Jul 29 2020 16:17:00.960]  LOG      params: {"+is_first_session":false,"+clicked_branch_link":false}
```

Reply:
Hmm that part does make sense to me but I tried using `reInitSession` in order to intra-app link into the currently open activity and here is what I got:

testing app:
testbed_simple

Actions:
App killed
Open with link, https://zjef.test-app.link/n4efBZnSu8 (see logs below at second 38) 
background and foreground (see logs below at second 51)
Link to the currently open activity see logs below at second 02)

Changes made to the code to enable intra app linking to the currently open Activity:
```
@ReactMethod
public void disableTracking(boolean disable) {
    Branch branch = Branch.getInstance();
    branch.disableTracking(disable);
    // https://zjef.test-app.link/n4efBZnSu8
    Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("https://zjef.test-app.link/n4efBZnSu8"));
    intent.putExtra("branch_force_new_session", true);
    mActivity.startActivity(intent);

    // Branch branch = Branch.getInstance();
    // branch.disableTracking(disable);
}


public void onNewIntent(Intent intent) {
	super.onNewIntent(intent);
	RNBranchModule.setIntent(intent, this);
	// setIntent(intent);
	RNBranchModule.reInitSession(this);
}
```
Observed logs following the "Actions" listed above:
```
[Fri Jul 31 2020 14:34:38.422]  INFO     Initializing react-native-branch v. 5.0.0
[Fri Jul 31 2020 14:34:38.423]  INFO     Subscribing to Branch links
[Fri Jul 31 2020 14:34:38.423]  LOG      Running "testbed_simple" with {"rootTag":1}
[Fri Jul 31 2020 14:34:38.570]  INFO     Received link response from Branch for rntestbed://open?link_click_id=817856439472920378
[Fri Jul 31 2020 14:34:38.571]  LOG      params: {"~channel":"Copy","~referring_link":"https://zjef.test-app.link/n4efBZnSu8","+is_first_session":false,"$one_time_use":false,"+clicked_branch_link":true,"$canonical_identifier":"abc","+match_guaranteed":true,"+click_timestamp":1596231276,"~creation_source":2,"$og_title":"wallo","~id":"816747731456192789","~feature":"share","$identity_id":"815317995815593579","source":"android","$publicly_indexable":"true"}

[Fri Jul 31 2020 14:34:51.695]  INFO     Received link response from Branch for null
[Fri Jul 31 2020 14:34:51.695]  LOG      params: {"+is_first_session":false,"+clicked_branch_link":false}

[Fri Jul 31 2020 14:35:02.708]  LOG      disableTracking Tracking Enabled
[Fri Jul 31 2020 14:35:02.757]  LOG      Branch opening URI https://zjef.test-app.link/n4efBZnSu8 
[Fri Jul 31 2020 14:35:02.849]  INFO     Received link response from Branch for null
[Fri Jul 31 2020 14:35:02.850]  LOG      params: {"~channel":"Copy","~referring_link":"https://zjef.test-app.link/n4efBZnSu8","+is_first_session":false,"$one_time_use":false,"+clicked_branch_link":true,"$canonical_identifier":"abc","+match_guaranteed":true,"+click_timestamp":1596231302,"~creation_source":2,"$og_title":"wallo","~id":"816747731456192789","~feature":"share","$identity_id":"815317995815593579","source":"android","$publicly_indexable":"true"}
```

So the only problem I see is this line `[Fri Jul 31 2020 14:34:51.695]  INFO     Received link response from Branch for null` - seems like it should give the link instead of a null in that intra-app linking scenario.

Reply:
@bklastaitis-branch Thanks for the testing, and sorry for the delay. I guess the purpose of doing this in disableTracking was just to reuse the first button in the testbed_simple app? There is a method exposed to JS to do this already: `branch.openURL(url, options)`. Here's the Android implementation: https://github.com/BranchMetrics/react-native-branch-deep-linking-attribution/blob/master/android/src/main/java/io/branch/rnbranch/RNBranchModule.java#L615. This differs from yours a little. It uses a specific Activity class (mActivity.getClass()) and adds the URL as a `branch` extra to the intent. It also allows a `newActivity: true` option that results in mActivity.finish() being called if present (Android only). This was an effort to deal with Android lifecycle issues in the past, but it doesn't make for a very good user experience.

Also at times we've had to recommend that users not call reInitSession because of certain lifecycle issues. I'd have to dig through Jira tickets to find what the issues were.

I added an openURL button to testbed_simple to exercise this functionality (along with the call to reInitSession in onNewIntent). It's at the bottom of the screen, and you may have to scroll. I mucked around with the RNBranchModule.openURL method a bit, rewriting it similarly to yours, but this is my result from the openURL button:

```
[Tue Aug 04 2020 11:15:06.816]  INFO     Initializing react-native-branch v. 5.0.0
[Tue Aug 04 2020 11:15:06.817]  INFO     Subscribing to Branch links
[Tue Aug 04 2020 11:15:06.821]  LOG      Running "testbed_simple" with {"rootTag":1}
[Tue Aug 04 2020 11:15:10.523]  LOG      Branch opening URI https://zjef.test-app.link/n4efBZnSu8 
```

The onOpenComplete callback is not called, only onOpenStart. Adding `{newActivity: true}` as an option to openURL makes matters worse. The onOpenStart method is not even called. I'm also testing on an emulator right now and will need to set up a device to test further.

The browser_example in this repo is intended to demonstrate intra-app linking. You can visit any page in a webview, share it using Branch, and then paste the link into the browser bar in the app to follow the link. I haven't yet updated it to RN 0.62.